mitodl/ocw-hugo-themes
# offline www theme

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-hugo-themes/issues/1111

#### What's this PR do?
This PR adds a new theme, `www-offline`. The purpose of this theme is to iterate content of type "website" and generate a paginated index of to be used in OCW mirror drives. It uses the same content as the online `www` theme and is intended to be used as a root website in the same way `www` is, but for offline use. Some of the commonly used partials in offline sites were also consolidated into `base-offline`, a theme which all offline sites can inherit to get baseline of how to handle URLs in an offline site. A new bundle, `www_offline` was added for this that only brings in CSS / JS needed for the offline index.

#### How should this be manually tested?
 - Clone `ocw-hugo-projects` and checkout the `cg/ocw-www-offline` branch
 - Clone the `ocw-www` content repo from `ocw-content-rc`: https://github.mit.edu/ocw-content-rc/ocw-www
 - Run `yarn build /path/to/ocw-content-rc/ocw-www /path/to/ocw-hugo-projects/ocw-www/config-offline.yaml`
 - In your `ocw-www` content repo, browse to the `dist` folder and open `index.html` in your browser, you should see a list of courses like in the screenshots
 - To test the links to courses, pick a course and download its offline build from https://live-qa.ocw.mit.edu. Extract the contents into a folder matching the URL to the course in the index page, which is generated using the `url_path` of the site. For example, `courses/1-963-a-sustainable-transportation-plan-for-mit-spring-2007` would be the path relative to the `dist` folder for https://live-qa.ocw.mit.edu/courses/1-963-a-sustainable-transportation-plan-for-mit-spring-2007/
 - After placing the relevant offline course site in your `dist/courses` folder, test out the link to the course in the index page
 - Test out the pagination controls at the bottom of the index page, ensuring that the buttons always work
 
 #### Where should the reviewer start?
I would start by looking at `www-offline/layouts/home.html`, which is where a bulk of the work is done. Also of interest will be `www-offline/layouts/partials/sort_websites_by_course_number.html`, which is required to do a "natural sort" on the courses by course number. Hugo only supports alphabetical / lexicographical sort, so by default it will put course numbers starting with 10 after course numbers starting with 1. This partial addresses that.

#### Any background context you want to provide?
This work is part of https://github.com/mitodl/hq/issues/785, an effort to generate OCW mirror drives that can be viewed offline or hosted off-site by a third party.

#### Out of scope
 - Currently the links in the header will point to the live site indicated by the env variable `STATIC_API_BASE_URL`. This is the same behavior as the offline course sites. Their primary use case is to be downloaded ad-hoc in ZIP archives from the live site, so the idea is if a user downloads a course and is browsing it, if they are online then the links in the header will go to the live site. We may want different behavior here for the mirror drives, but that will be addressed in a different PR.
 - There is a known issue with loading the MaterialIcons font. The path is incorrect on pages other than the home page. The `relativeUrls` setting is being used so the site can be viewed offline, but this setting does not work on CSS `url` properties currently as noted here: https://github.com/gohugoio/hugo/issues/4355
 
 #### Screenshots
 
![image](https://user-images.githubusercontent.com/12089658/227042869-4242e8d9-5405-4dcb-b3ac-c43af79baa57.png)
![image](https://user-images.githubusercontent.com/12089658/227043530-53f5c552-135d-4814-ac29-57c20eeccbcc.png)
